src/http/ls.js

```diff
@@ -14,103 +14,83 @@ const mapEntry = (entry) => {
   }
 }
 
-const mfsLs = (api) => {
-  api.route({
-    method: 'POST',
-    path: '/api/v0/files/ls',
-    config: {
-      handler: (request, reply) => {
-        const {
-          ipfs
-        } = request.server.app
-        const {
-          arg,
+const mfsLs = {
+  method: 'POST',
+  path: '/api/v0/files/ls',
+  async handler (request, h) {
+    const {
+      ipfs
+    } = request.server.app
+    const {
+      arg,
+      long,
+      cidBase,
+      stream
+    } = request.query
+
+    if (stream) {
+      const responseStream = await new Promise((resolve, reject) => {
+        const readableStream = ipfs.files.lsReadableStream(arg, {
           long,
-          cidBase,
-          stream
-        } = request.query
+          cidBase
+        })
 
-        if (stream) {
-          const readableStream = ipfs.files.lsReadableStream(arg, {
-            long,
-            cidBase
-          })
+        let passThrough
 
-          if (!readableStream._read) {
-            // make the stream look like a Streams2 to appease Hapi
-            readableStream._read = () => {}
-            readableStream._readableState = {}
+        readableStream.on('data', (entry) => {
+          if (!passThrough) {
+            passThrough = new PassThrough()
+            resolve(passThrough)
           }
 
-          let passThrough
-
-          readableStream.on('data', (entry) => {
-            if (!passThrough) {
-              passThrough = new PassThrough()
-
-              reply(passThrough)
-                .header('X-Stream-Output', '1')
-            }
+          passThrough.write(JSON.stringify(mapEntry(entry)) + '\n')
+        })
 
-            passThrough.write(JSON.stringify(mapEntry(entry)) + '\n')
-          })
+        readableStream.once('end', (entry) => {
+          if (passThrough) {
+            passThrough.end(entry ? JSON.stringify(mapEntry(entry)) + '\n' : undefined)
+          }
+        })
 
-          readableStream.once('end', (entry) => {
-            if (passThrough) {
-              passThrough.end(entry ? JSON.stringify(mapEntry(entry)) + '\n' : undefined)
-            }
-          })
+        readableStream.once('error', (error) => {
+          reject(error)
```

Comment:
What would happen here if an error occurred during writing (so we'd already resolved the promise)?

Reply:
A promise can only be resolved/rejected once, and first one wins, so nothing really. Hapi would have already started streaming to the client. At that point we've handed over responsibility of responding (by passing the stream to hapi) and my expectation would be for either Hapi or Node.js to handle an error in the stream and close the connection...

Previously we'd have called `reply` again, which definitely wouldn't have worked as the HTTP headers would have already been sent and possibly some data.

Reply:
We should probably at least log the error then as presumably it would get swallowed.

Reply:
Happy to do this later though.